File: topoflow/policy_graph.py
# SPDX-License-Identifier: AGPL-3.0-only
# Copyright (C) 2026 Discover Legal
"""Policy graph + UCB1 router (spec §6) [AF].

A tabular contextual bandit: one Edge per (Signature, Action) cell. UCB1 with
annealed exploration and a reliability (failure-rate) penalty. Backups are
trajectory-level and support confidence-weighted fractional visits.
"""
from __future__ import annotations

import json
from dataclasses import asdict, dataclass
from math import log, sqrt
from typing import Optional

from .config import Config
from .types import Action, Regime, Signature


@dataclass
class Edge:  # one (signature, action) cell
    visits: float = 0.0
    mean_reward: float = 0.0
    m2: float = 0.0  # for online variance (Welford)
    tokens_sum: int = 0
    failures: float = 0.0  # validation + recoverable-exec failures

    @property
    def variance(self) -> float:
        return self.m2 / self.visits if self.visits > 1 else 0.0

    @property
    def fail_rate(self) -> float:
        return self.failures / self.visits if self.visits > 0 else 0.0


class PolicyGraph:
    def __init__(self, cfg: Config):
        self.edges: dict[tuple[Signature, tuple], Edge] = {}
        self.sig_visits: dict[Signature, int] = {}
        self.cfg = cfg

    # ── scoring / selection ────────────────────────────────────────────────
    def _cs(self, ns: int) -> float:
        """Annealed exploration constant, eq (5)."""
        return max(self.cfg.ucb_floor, self.cfg.ucb_c0 * 2 ** (-ns / self.cfg.ucb_half_life))

    def score(self, sig: Signature, action: Action) -> float:  # eq (4)+(5)
        e = self.edges.get((sig, action.key()))
        if e is None or e.visits == 0:
            return float("inf")  # force initial exploration [AF]
        ns = self.sig_visits.get(sig, 0)
        ucb = self._cs(ns) * sqrt(log(ns + 1) / e.visits)
        return e.mean_reward + ucb - self.cfg.lam * e.fail_rate

    def select(self, sig: Signature, legal: list[Action]) -> Action:
        if not legal:
            raise ValueError("select() called with no legal actions")
        return max(legal, key=lambda a: self.score(sig, a))

    # ── backup ─────────────────────────────────────────────────────────────
    def backup(
        self,
        sig: Signature,
        action: Action,
        reward: float,
        tokens: int,
        failed: bool,
        weight: float = 1.0,
    ) -> None:
        """Trajectory-level backup [AF].

        weight < 1 implements confidence-weighted updates (low-confidence judge
        rewards count as a fractional visit). weight == 1 for verifiable Q.
        """
        if weight <= 0:
            return
        key = (sig, action.key())
        e = self.edges.setdefault(key, Edge())
        delta = reward - e.mean_reward
        e.visits += weight
        e.mean_reward += (weight / e.visits) * delta
        e.m2 += weight * delta * (reward - e.mean_reward)
        e.tokens_sum += tokens
        e.failures += weight if failed else 0.0
        self.sig_visits[sig] = self.sig_visits.get(sig, 0) + 1

    # ── persistence ────────────────────────────────────────────────────────
    def save(self, path: str) -> None:
        records = []
        for (sig, akey), e in self.edges.items():
            records.append(
                {
                    "sig": _sig_to_json(sig),
                    "action_key": list(akey),
                    "edge": asdict(e),
                }
            )
        blob = {
            "sig_visits": [
                {"sig": _sig_to_json(s), "visits": v} for s, v in self.sig_visits.items()
            ],
            "edges": records,
        }
        with open(path, "w") as f:
            json.dump(blob, f, indent=2)

    @classmethod
    def load(cls, path: str, cfg: Config) -> "PolicyGraph":
        with open(path) as f:
            blob = json.load(f)
        g = cls(cfg)
        for rec in blob["edges"]:
            sig = _sig_from_json(rec["sig"])
            akey = tuple(rec["action_key"])
            g.edges[(sig, akey)] = Edge(**rec["edge"])
        for sv in blob["sig_visits"]:
            g.sig_visits[_sig_from_json(sv["sig"])] = sv["visits"]
        return g

    # ── inspection ─────────────────────────────────────────────────────────
    def summary(self) -> dict:
        """Human-inspectable dump: per signature, actions with stats."""
        out: dict[str, list[dict]] = {}
        for (sig, akey), e in self.edges.items():
            out.setdefault(_sig_label(sig), []).append(
                {
                    "action": list(akey),
                    "visits": round(e.visits, 3),
                    "mean_reward": round(e.mean_reward, 4),
                    "variance": round(e.variance, 4),
                    "fail_rate": round(e.fail_rate, 4),
                    "tokens_sum": e.tokens_sum,
                }
            )
        for label in out:
            out[label].sort(key=lambda r: r["mean_reward"], reverse=True)
        return out


# ── (de)serialization helpers ──────────────────────────────────────────────
def _sig_to_json(sig: Signature) -> dict:
    return {
        "regime": sig.regime.value,
        "handoff_mask": list(sig.handoff_mask),
        "correctness_b": sig.correctness_b,
        "uncertainty_b": sig.uncertainty_b,
        "contradiction_b": sig.contradiction_b,
        "evidence_b": sig.evidence_b,
    }


def _sig_from_json(d: dict) -> Signature:
    return Signature(
        regime=Regime(d["regime"]),
        handoff_mask=tuple(d["handoff_mask"]),
        correctness_b=d["correctness_b"],
        uncertainty_b=d["uncertainty_b"],
        contradiction_b=d["contradiction_b"],
        evidence_b=d["evidence_b"],
    )


def _sig_label(sig: Signature) -> str:
    return (
        f"{sig.regime.value}|m{''.join(map(str, sig.handoff_mask))}"
        f"|c{sig.correctness_b}u{sig.uncertainty_b}"
        f"k{sig.contradiction_b}e{sig.evidence_b}"
    )

File: biglaw-go/internal/topoflow/policygraph.go
// SPDX-License-Identifier: AGPL-3.0-only
// Copyright (C) 2026 Discover Legal

package topoflow

import (
	"encoding/json"
	"math"
	"os"
	"sort"
)

// Edge is one (signature, action) cell [AF].
type Edge struct {
	Visits     float64 `json:"visits"`
	MeanReward float64 `json:"meanReward"`
	M2         float64 `json:"m2"` // online variance (Welford)
	TokensSum  int     `json:"tokensSum"`
	Failures   float64 `json:"failures"`
}

// Variance returns the online population variance.
func (e *Edge) Variance() float64 {
	if e.Visits > 1 {
		return e.M2 / e.Visits
	}
	return 0
}

// FailRate returns failures / visits.
func (e *Edge) FailRate() float64 {
	if e.Visits > 0 {
		return e.Failures / e.Visits
	}
	return 0
}

type edgeKey struct {
	Sig Signature
	Act Action
}

// PolicyGraph is a tabular contextual bandit over (signature, action) cells.
type PolicyGraph struct {
	edges     map[edgeKey]*Edge
	sigVisits map[Signature]int
	cfg       Config
}

// NewPolicyGraph creates an empty policy graph.
func NewPolicyGraph(cfg Config) *PolicyGraph {
	return &PolicyGraph{edges: map[edgeKey]*Edge{}, sigVisits: map[Signature]int{}, cfg: cfg}
}

func (g *PolicyGraph) cs(ns int) float64 {
	return math.Max(g.cfg.UCBFloor, g.cfg.UCBc0*math.Pow(2, -float64(ns)/float64(g.cfg.UCBHalfLife)))
}

// Score implements eq (4)+(5). Unvisited actions return +Inf to force exploration.
func (g *PolicyGraph) Score(sig Signature, a Action) float64 {
	e := g.edges[edgeKey{sig, a}]
	if e == nil || e.Visits == 0 {
		return math.Inf(1)
	}
	ns := g.sigVisits[sig]
	ucb := g.cs(ns) * math.Sqrt(math.Log(float64(ns)+1)/e.Visits)
	return e.MeanReward + ucb - g.cfg.Lam*e.FailRate()
}

// Select returns the highest-scoring legal action (first on ties).
func (g *PolicyGraph) Select(sig Signature, legal []Action) Action {
	best := legal[0]
	bestScore := g.Score(sig, best)
	for _, a := range legal[1:] {
		if s := g.Score(sig, a); s > bestScore {
			best, bestScore = a, s
		}
	}
	return best
}

// Backup performs a trajectory-level update [AF]. weight < 1 implements
// confidence-weighted backups (fractional visits); weight == 1 for verifiable Q.
func (g *PolicyGraph) Backup(sig Signature, a Action, reward float64, tokens int, failed bool, weight float64) {
	if weight <= 0 {
		return
	}
	k := edgeKey{sig, a}
	e := g.edges[k]
	if e == nil {
		e = &Edge{}
		g.edges[k] = e
	}
	delta := reward - e.MeanReward
	e.Visits += weight
	e.MeanReward += (weight / e.Visits) * delta
	e.M2 += weight * delta * (reward - e.MeanReward)
	e.TokensSum += tokens
	if failed {
		e.Failures += weight
	}
	g.sigVisits[sig]++
}

// ── persistence ─────────────────────────────────────────────────────────────

type edgeRecord struct {
	Sig  Signature `json:"sig"`
	Act  Action    `json:"act"`
	Edge Edge      `json:"edge"`
}

type sigVisitRecord struct {
	Sig    Signature `json:"sig"`
	Visits int       `json:"visits"`
}

type graphBlob struct {
	Edges     []edgeRecord     `json:"edges"`
	SigVisits []sigVisitRecord `json:"sigVisits"`
}

// Save writes the graph to a JSON file (the warm-start artifact).
func (g *PolicyGraph) Save(path string) error {
	var blob graphBlob
	for k, e := range g.edges {
		blob.Edges = append(blob.Edges, edgeRecord{Sig: k.Sig, Act: k.Act, Edge: *e})
	}
	for s, v := range g.sigVisits {
		blob.SigVisits = append(blob.SigVisits, sigVisitRecord{Sig: s, Visits: v})
	}
	data, err := json.MarshalIndent(blob, "", "  ")
	if err != nil {
		return err
	}
	return os.WriteFile(path, data, 0o644)
}

// LoadPolicyGraph reads a graph from a JSON file.
func LoadPolicyGraph(path string, cfg Config) (*PolicyGraph, error) {
	data, err := os.ReadFile(path)
	if err != nil {
		return nil, err
	}
	var blob graphBlob
	if err := json.Unmarshal(data, &blob); err != nil {
		return nil, err
	}
	g := NewPolicyGraph(cfg)
	for _, r := range blob.Edges {
		e := r.Edge
		g.edges[edgeKey{r.Sig, r.Act}] = &e
	}
	for _, r := range blob.SigVisits {
		g.sigVisits[r.Sig] = r.Visits
	}
	return g, nil
}

// Clone returns a deep copy (used for warm-start arms).
func (g *PolicyGraph) Clone() *PolicyGraph {
	c := NewPolicyGraph(g.cfg)
	for k, e := range g.edges {
		ec := *e
		c.edges[k] = &ec
	}
	for s, v := range g.sigVisits {
		c.sigVisits[s] = v
	}
	return c
}

// Summary returns a human-inspectable per-signature view, best action first.
func (g *PolicyGraph) Summary() map[string][]map[string]any {
	out := map[string][]map[string]any{}
	for k, e := range g.edges {
		label := sigLabel(k.Sig)
		out[label] = append(out[label], map[string]any{
			"action":     k.Act,
			"visits":     e.Visits,
			"meanReward": e.MeanReward,
			"variance":   e.Variance(),
			"failRate":   e.FailRate(),
			"tokensSum":  e.TokensSum,
		})
	}
	for label := range out {
		rows := out[label]
		sort.Slice(rows, func(i, j int) bool {
			return rows[i]["meanReward"].(float64) > rows[j]["meanReward"].(float64)
		})
	}
	return out
}

func sigLabel(s Signature) string {
	digits := func(a [7]int) string {
		b := make([]byte, 7)
		for i, v := range a {
			b[i] = byte('0' + v)
		}
		return string(b)
	}
	return string(s.Regime) + "|m" + digits(s.Mask)
}
